File: src/tree/cci/MinimalTree.java
package tree.cci;

/* Minimal Tree: Given a sorted (increasing order) array with unique integer elements,
write an algorithm to create a binary search tree with minimal height.
 */
public class MinimalTree {
    class Node {
        int data;
        Node left, right;

        Node(int data){
            this.data = data;
            left = right = null;
        }
    }

    Node createMinimalBST(int[] arr) {
        return createMinimalBST(arr, 0, arr.length -1);
    }

    Node createMinimalBST(int[] arr, int start, int end) {
        if (end < start) {
            return null;
        }
        int mid = (start + end) / 2;
        Node n = new Node(arr[mid]);

        n.left = createMinimalBST(arr, start, mid - 1);
        n.right = createMinimalBST(arr, mid + 1, end);
        return n;
    }

    public static void main(String args[]) {
        MinimalTree mt = new MinimalTree();

        int a[] = {2, 7, 4, 1, 5, 3};
        Node root = mt.createMinimalBST(a);
        System.out.println("Done!");
    }
}

File: node/cracking-the-coding-interview/src/treengraphs/MinimalTree.js
class Node {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class BinarySearchTree {
    constructor() {
        this.root = null;
    }

    createMinimalBST(arr) {
        return this.createMinimalBSTRe(arr, 0, arr.length - 1);
    }

    createMinimalBSTRe(arr, start, end) {
        if (end < start) {
            return null;
        }

        let mid = Math.floor((start + end) / 2);
        const n = new Node(arr[mid]);
        n.left = this.createMinimalBSTRe(arr, start, mid - 1);
        n.right = this.createMinimalBSTRe(arr, mid + 1, end);
        return n;
    }
}

const bst = new BinarySearchTree();

// int a[] = {2, 7, 4, 1, 5, 3};
const a = [1, 2, 3, 4, 5, 7];
const root = bst.createMinimalBST(a);
console.log("Done!");

